pkg/app/piped/executor/kubernetes/kubernetes.go

```diff
@@ -605,3 +606,126 @@ func annotateConfigHashToDeployment(m provider.Manifest, managedConfigMaps, mana
 	)
 	return nil
 }
+
+type patcher func(m provider.Manifest, cfg config.K8sResourcePatch) (*provider.Manifest, error)
+
+func patchManifests(manifests []provider.Manifest, patches []config.K8sResourcePatch, patcher patcher) ([]provider.Manifest, error) {
+	if len(patches) == 0 {
+		return manifests, nil
+	}
+
+	out := make([]provider.Manifest, len(manifests))
+	copy(out, manifests)
+
+	for _, p := range patches {
+		target := -1
+		for i, m := range out {
+			if m.Key.Kind != p.Target.Kind {
+				continue
+			}
+			if m.Key.Name != p.Target.Name {
+				continue
+			}
+			target = i
+			break
+		}
+		if target < 0 {
+			return nil, fmt.Errorf("no manifest matches the given patch: kind=%s, name=%s", p.Target.Kind, p.Target.Name)
+		}
+		patched, err := patcher(out[target], p)
+		if err != nil {
+			return nil, fmt.Errorf("failed to patch manifest: %s, error: %v", out[target].Key, err)
+		}
+		out[target] = *patched
+	}
+
+	return out, nil
+}
+
+func patchManifest(m provider.Manifest, patch config.K8sResourcePatch) (*provider.Manifest, error) {
+	if len(patch.Ops) == 0 {
+		return &m, nil
+	}
+
+	fullBytes, err := m.YamlBytes()
+	if err != nil {
+		return nil, err
+	}
+
+	process := func(bytes []byte) ([]byte, error) {
+		p, err := yamlprocessor.NewProcessor(bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, o := range patch.Ops {
+			switch o.Op {
+			case config.K8sResourcePatchOpYAMLReplace:
+				if err := p.ReplaceString(o.Path, o.Value); err != nil {
+					return nil, fmt.Errorf("failed to replace value at path: %s, error: %v", o.Path, err)
+				}
+			default:
+				// TODO: Support more patch operation for K8sCanaryRolloutStageOptions.
+				return nil, fmt.Errorf("%s operation is not supported currently", o.Op)
+			}
+		}
+
+		return p.Bytes()
+	}
+
+	buildManifest := func(bytes []byte) (*provider.Manifest, error) {
+		manifests, err := provider.ParseManifests(string(bytes))
+		if err != nil {
+			return nil, err
+		}
+		if len(manifests) != 1 {
+			return nil, fmt.Errorf("unexpected number of manifests, expected 1, got %d", len(manifests))
+		}
+		return &manifests[0], nil
+	}
+
+	// When the target is the whole manifest,
+	// just pass full bytes to process and build a new manifest based on the returned data.
+	root := patch.Target.DocumentRoot
+	if root == "" {
+		out, err := process(fullBytes)
+		if err != nil {
+			return nil, err
+		}
+		return buildManifest(out)
+	}
+
+	// When the target is a single string field
+	// we have to extract the field value as a string data firstly.
```

Comment:
```suggestion
	// When the target is a manifest field,
	// we have to extract that field value as a string.
```
nit, but it's up to you 👌